src/index.js

```diff
@@ -13,45 +13,94 @@
  * @returns {typeof EPromise} EPromise class
  */
 export default function makeEPromiseClass(BasePromise) {
-  // A remoteRelay must additionally have an AWAIT_FAR method
-  const localRelay = {
-    GET(p, key) {
-      return p.then(o => o[key]);
-    },
-    PUT(p, key, val) {
-      return p.then(o => (o[key] = val));
-    },
-    DELETE(p, key) {
-      return p.then(o => delete o[key]);
-    },
-    POST(p, optKey, args) {
-      if (optKey === undefined || optKey === null) {
-        return p.then(o => o(...args));
-      }
-      return p.then(o => o[optKey](...args));
-    },
-  };
-
-  const relayToPromise = new WeakMap();
+  const presenceToResolvedRelay = new WeakMap();
   const promiseToRelay = new WeakMap();
 
+  // This special relay accepts Promises, and forwards
+  // the resolved Presence to its corresponding resolvedRelay.
+  //
+  // If passed a Promise that is not a Presence, perform
+  // the corresponding local operation.
+  let forwardingRelay;
   function relay(p) {
-    return promiseToRelay.get(p) || localRelay;
+    return promiseToRelay.get(p) || forwardingRelay;
   }
 
   class EPromise extends BasePromise {
-    static makeRemote(remoteRelay) {
-      const promise = this.resolve(remoteRelay.AWAIT_FAR());
-      relayToPromise.set(remoteRelay, promise);
-      promiseToRelay.set(promise, remoteRelay);
-      return promise;
+    static makePresence(executor, unresolvedRelay) {
+      let presenceResolve;
+      let presenceReject;
+      let relayResolve;
+      let relayReject;
+      const presenceP = new EPromise((resolve, reject) => {
+        presenceResolve = resolve;
+        presenceReject = reject;
+      });
+      const relayP = new EPromise((resolve, reject) => {
+        relayResolve = resolve;
+        relayReject = reject;
+      });
+
+      function postpone(forwarded) {
+        // Just wait until the relay is resolved/rejected.
+        return async (p, ...args) => {
+          await relayP;
+          return p[forwarded](args);
+        };
+      }
+
+      if (!unresolvedRelay) {
+        // Create a simple unresolvedRelay that just postpones until the
+        // resolvedRelay is set.
+        //
+        // This is insufficient for remote Presences (too many round-trips),
+        // but is an easy way to create a local Presence.
+        unresolvedRelay = {
+          GET: postpone('get'),
+          PUT: postpone('put'),
+          DELETE: postpone('del'),
+          POST: postpone('post'),
+        };
+      }
+
+      // Until the presence is resolved, we use the unresolvedRelay.
+      promiseToRelay.set(presenceP, unresolvedRelay);
+
+      function resolvePresence(presence, resolvedRelay) {
+        if (resolvedRelay) {
+          // FIXME: What if the presence is already in the table?
```

Comment:
The error you have is good. But must also check that the alleged presence is a non-promise and non-thenable. If the resolvedRelay is absent, then the `presence` argument could be anything that we could normally provide the resolve.

Reply:
Fixed in next commit, and more sanity checks added.